File: starters/flutter/lib/src/core/session/session_hook.dart
const sessionHookContractVersion = 'session-hook/2.0.0';

enum SessionState { unknown, anonymous, authenticated }

abstract interface class SessionHook {
  String get contractVersion;
  SessionState current();
  void Function() subscribe(void Function(SessionState state) listener);
}

final class AnonymousSessionHook implements SessionHook {
  const AnonymousSessionHook();

  @override
  String get contractVersion => sessionHookContractVersion;

  @override
  SessionState current() => SessionState.anonymous;

  @override
  void Function() subscribe(void Function(SessionState state) listener) {
    return () {};
  }
}

File: starters/react-native/src/session/session-hook.ts
export const SESSION_HOOK_CONTRACT_VERSION = 'session-hook/2.0.0';

export type SessionState = 'unknown' | 'anonymous' | 'authenticated';

export interface SessionHook {
  readonly contractVersion: typeof SESSION_HOOK_CONTRACT_VERSION;
  current(): SessionState;
  subscribe(listener: (state: SessionState) => void): () => void;
}

export function createAnonymousSessionHook(): SessionHook {
  return Object.freeze({
    contractVersion: SESSION_HOOK_CONTRACT_VERSION,
    current: () => 'anonymous',
    subscribe: () => () => undefined,
  });
}
